docs/proposals/20210310-edge-device-management.md

```diff
@@ -0,0 +1,498 @@
+    Addressable Addressable `json:"addressable,omitempty"`
+}
+
+// DeviceServiceStatus defines the observed state of DeviceService
+type DeviceServiceStatus struct {
+    // ServiceId is the Id assigned by the EdgeX foundry
+    ServiceId string `json:"id,omitempty"`
+    // The condition of the service
+    Condition DeviceServiceCondition `json:"condition,omitempty"`
+}
+
+type DeviceServiceCondition string
+
+const (
+	Unavailable DeviceServiceCondition = "Unavailable"
+    Available DeviceServiceCondition = "Availale"
+)
+
+type Addressable struct {
+    // ID is a unique identifier for the Addressable, such as a UUID
+    Id string `json:"id,omitempty"`
+    // Name is a unique name given to the Addressable
+    Name string `json:"name,omitempty"`
+    // Protocol for the address (HTTP/TCP)
+    Protocol string `json:"protocol,omitempty"`
+    // Address of the addressable
+    Address string `json:"address,omitempty"`
+    // Port for the address
+    Port int `json:"port,omitempty"`
+    // Path for callbacks
+    Path string `json:"path,omitempty"`
+    // For message bus protocols
+    Publisher string `json:"publisher,omitempty"`
+    // Topic for message bus addressables
+    Topic string `json:"topic,omitempty"`
+}
+```
+
+### Interaction with EdgeX Foundry
+
+#### 1. Setting up EdgeX Foundry
+
+As OpenYurt supports all upstream Kubernetes features, the EdgeX Foundry deployment should be the same as deploying it on the Kubernetes. The Deployment yaml files can be found in the [EdgeX Foundry example repository](https://github.com/edgexfoundry/edgex-examples/tree/master/deployment/kubernetes), by simply applying them, the necessary services will be deployed. Also, we will allow users to choose if they want to set up the EdgeX Foundry automatically when converting the Kubernetes cluster to the OpenYurt cluster using the `yurtctl` command-line tool.
+
+#### 2. CRUD objects on EdgeX Foundry
+
+To conciliate the device states between OpenYurt and EdgeX Foundry, a new component, DeviceManager, will be installed on the OpenYurt. The DeviceManager includes three controllers, i.e., DeviceProfile Controller, DeviceService Controller, and the Device Controller, which act as mediators between the OpenYurt and the EdgeX Foundry and are responsible for reconciling the states of the device-related custom resources with the states of the corresponding objects on the EdgeX Foundry.
+
+Following is the process of connecting a new device to the OpenYurt through EdgeX Foundry.
+
+![connect-device](../img/connect-device.png)
+
+The vendor applies the DeviceProfile CR.
+2. The DeviceManager creates the related ValueDescriptor on the EdgeX MetaData Service.
+3. The DeviceManager create the DeviceProfile on the EdgeX CoreData Service.
+4. The end-user applies the Device CR.
+5. The DeviceManager creates the Device object on the EdgeX CoreData Service.
+6. The Device supported commands are created on the EdgeX Command Service.
```

Comment:
The command service is stateless, it doesn't store any data, it will read data from the metadata service and then do the command translation if I remember correctly .

Reply:
Aha, I see. Thanks for clarifying.